src/aspire/source/relion.py

```diff
@@ -119,26 +192,63 @@ def __init__(self, filepath, pixel_size=1, B=0, n_workers=-1, ignore_missing_fil
                     voltage=row[0],
                     defocus_u=row[1],
                     defocus_v=row[2],
-                    defocus_ang=row[3],
+                    defocus_ang=row[3] * np.pi / 180,  # degrees to radians
                     Cs=row[4],
                     alpha=row[5],
-                    B=self.B
+                    B=B
                 )
             )
-        filters = SourceFilter(filters, indices=filter_indices)
 
-        offsets = self.df[['rlnOriginX', 'rlnOriginY']].values
-        amplitudes = np.ones(self.n)
-        states = self.df['rlnClassNumber'].values
+        metadata['__filter'] = [filters[i] for i in filter_indices]
 
         ImageSource.__init__(
             self,
-            L=self.L,
-            n=self.n,
-            states=states,
-            filters=filters,
-            offsets=offsets,
-            amplitudes=amplitudes,
-            rots=rots,
-            dtype=self.dtype
+            L=L,
+            n=n,
+            dtype=dtype,
+            metadata=metadata
         )
+
+    def __str__(self):
+        return f'Starfile ({self.n} images of size {self.L}x{self.L})'
+
+    def _images(self, start=0, num=None):
+
+        def load_single_mrcs(filepath, df):
+            arr = mrcfile.open(filepath).data
+            data = arr[df['__mrc_index'] - 1, :, :].T
+
+            if self.L < self._L:
```

Comment:
This is confusing. What's `self.L` and what's `self._L`? Maybe need more descriptive names.

Reply:
Yes. This is an issue that is best handled by the upcoming `Transformer` class. Opened an issue #46 to keep track of this. Closing here.